select for update statement can't got stmt-count-limit error

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

For minimum reproduce, the TiDB's configuration as follow:

```sql
[performance]
stmt-count-limit=3
```

Reproduce SQL:

```sql
create table t (id int key);
set @@tidb_disable_txn_auto_retry=0;
set autocommit=0;
begin optimistic;
insert into t values (1);
insert into t values (2);
select * from t for update;
insert into t values (3);
insert into t values (4);
commit;
select * from t;
```


<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

When execute `select * from t for update;`, should got error:

```sql
mysql> set @@tidb_disable_txn_auto_retry=0;
Query OK, 0 rows affected (0.00 sec)

mysql> set autocommit=0;
Query OK, 0 rows affected (0.00 sec)

mysql> begin optimistic;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t values (2);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t for update;
ERROR 1105 (HY000): statement count 4 exceeds the transaction limitation, transaction has been rollback, autocommit = false
mysql> insert into t values (3);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t values (4);
Query OK, 1 row affected (0.00 sec)

mysql> commit;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+----+
| id |
+----+
|  3 |
|  4 |
+----+
2 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)

All SQL are executed successfully, but the finally select result are following, **the user may think that the transaction has been partially committed.**

```
mysql> set @@tidb_disable_txn_auto_retry=0;
Query OK, 0 rows affected (0.00 sec)

mysql> set autocommit=0;
Query OK, 0 rows affected (0.00 sec)

mysql> begin optimistic;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t values (2);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t for update;
+----+
| id |
+----+
|  1 |
|  2 |
+----+
2 rows in set (0.00 sec)

mysql> insert into t values (3);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t values (4);
Query OK, 1 row affected (0.00 sec)

mysql> commit;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+----+
| id |
+----+
|  3 |
|  4 |
+----+
2 rows in set (0.00 sec)
```

But you can find an error log in TiDB log file:

```
[2023/11/08 16:13:29.486 +08:00] [ERROR] [terror.go:317] ["function call errored"] [error="statement count 4 exceeds the transaction limitation, autocommit = false"] [stack="github.com/pingcap/tidb/pkg/parser/terror.Call\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/parser/terror/terror.go:317\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleStmt\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/server/conn.go:2042\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/server/conn.go:1800\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).dispatch\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/server/conn.go:1287\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).Run\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/server/conn.go:1066\ngithub.com/pingcap/tidb/pkg/server.(*Server).onConn\n\t/Users/cs/code/goread/src/github.com/pingcap/tidb/pkg/server/server.go:701"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master: 89ec3d38b2c26b45b32208ef73b57461ea59908e

--- a/pkg/server/conn.go
+++ b/pkg/server/conn.go
@@ -2049,11 +2049,7 @@ func (cc *clientConn) handleStmt(ctx context.Context, stmt ast.StmtNode, warns [
 			execStmt.(*executor.ExecStmt).FinishExecuteStmt(0, err, false)
 		}
 	}
-	if err != nil {
-		return false, err
-	}
-
-	return false, nil
+	return false, err
 }
 
 func (cc *clientConn) handleFileTransInConn(ctx context.Context, status uint16) (bool, error) {
@@ -2284,6 +2280,9 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs resultset.ResultSet, b
 		if stmtDetail != nil {
 			stmtDetail.WriteSQLRespDuration += time.Since(start)
 		}
+	}
+	if err := rs.Close(); err != nil {
+		return false, err
 	}
 
 	if stmtDetail != nil {

--- a/pkg/session/session.go
+++ b/pkg/session/session.go
@@ -2467,16 +2467,20 @@ const ExecStmtVarKey ExecStmtVarKeyType = 0
 // RecordSet, so this struct exists and RecordSet.Close() is overrided handle that.
 type execStmtResult struct {
 	sqlexec.RecordSet
-	se  *session
-	sql sqlexec.Statement
+	se     *session
+	sql    sqlexec.Statement
+	closed bool
 }
 
 func (rs *execStmtResult) Close() error {
-	se := rs.se
-	if err := rs.RecordSet.Close(); err != nil {
-		return finishStmt(context.Background(), se, err, rs.sql)
+	if rs.closed {
+		return nil
 	}
-	return finishStmt(context.Background(), se, nil, rs.sql)
+	se := rs.se
+	err := rs.RecordSet.Close()
+	err = finishStmt(context.Background(), se, err, rs.sql)
+	rs.closed = true
+	return err
 }
 
 // rollbackOnError makes sure the next statement starts a new transaction with the latest InfoSchema.

